### ISSUE TITLE
Reorganize Project Structure for Improved Maintainability

### ISSUE THREAD

**Issue by dwikler** [sjswerdloff/tdwii_plus_examples#85]
As the project has grown and evolved, its structure has become a bit mixed. The current structure combines CLI tools, UI applications, and libraries within one folder. As a result, some files are using the same filename in subfolders, with similar goal but different contents, some configuration files are duplicated. Progressively restructuring the project with separation of concern would improve maintainability and extensibility.

**Comment by dwikler** [comment]
Here are 2 options we can discuss:

### option 1
```
tdwii-plus-examples/
├── cli/
│   ├── wrklstmgr.py
│   └── watchwrkitem.py
├── apps/
│   ├── rtbdi_creator.py
│   └── tdd.py
├── config/
│   ├── default.ini
│   └── known_ae.json
├── doc/
│   ├── cli.md
│   └── apps.md
├── tdwii_plus_examples/
│   ├── __init__.py
│   ├── basescp.py
│   └── echoscp.py
├── pyproject.toml
└── README.md
```
### option 2
```
tdwii-plus-examples/
├── tdwii_plus_examples/
│   ├── __init__.py
│   ├── basescp.py
│   ├── echoscp.py
│   ├── cli/
│   │   ├── wrklstmgr.py
│   │   └── watchwrkitem.py
│   ├── apps/
│   │   ├── rtbdi_creator.py
│   │   └── tdd.py
│   ├── config/
│   │   ├── default.ini
│   │   └── known_ae.json
│   └── doc/
│       ├── cli.md
│       └── apps.md
├── pyproject.toml
└── README.md
```

Notes: Poetry usually uses underscores for package folder when project name contains dashes

**Comment by sjswerdloff** [comment]
I agree that the file structure needs reorganising.  I prefer option 2 vs option 1, but I think having a library directory that contains the reusable (and already reused) modules would be helpful.

**Comment by dwikler** [comment]
Let's go for option 2 then 👍 I can start by duplicating the BaseSCP class from TDWII_PPVS_subscriber to the tdwii_plus_examples package folder and refactor the apps to use it, then make my way down the hierarchy of SCP classes.

**Comment by dwikler** [comment]
I have created the pull request  (#86) that refactors BaseSCP. I will continue with its subclass EchoSCP.

**Comment by dwikler** [comment]
The previous pull request (https://github.com/sjswerdloff/tdwii_plus_examples/pull/87) refactored EchoSCP in CEchoSCP

A new pull request (https://github.com/sjswerdloff/tdwii_plus_examples/pull/88) refactors its subclass StoreSCP in CStoreSCP
The next step is to refactor its subclass NEventReceiver

**Comment by dwikler** [comment]
A new pull request (https://github.com/sjswerdloff/tdwii_plus_examples/pull/91) refactors NEventReceiver class in UPSNEventReceiver and PPVS_SCP class

**Comment by dwikler** [comment]
@sjswerdloff I would need advices on the next step, the refactoring of upsscp.py.
It is not currently a class and I was thinking to refactor it in a class similar to the SCP and Receiver classes and a CLI app.
The question is whether the class should include all presentation contexts which are currently:
UPS Push with N-CREATE primitive from the UPS Push Service group,
UPS Pull with C-FIND, N-ACTION - Change UPS State, N-GET, N-SET primitives from the UPS Pull Service group,
or rather create distinct class for each SOP class (where the combination would be handled in another class or directly in the CLI app.
As the primitive C-FIND, N-SET and N-GET are present in multiple UPS service groups, one handler class for all DIMSE primitive makes sense. What do you think?

**Comment by sjswerdloff** [comment]
> @sjswerdloff I would need advices on the next step, the refactoring of upsscp.py. It is not currently a class and I was thinking to refactor it in a class similar to the SCP and Receiver classes and a CLI app. The question is whether the class should include all presentation contexts which are currently: UPS Push with N-CREATE primitive from the UPS Push Service group, UPS Pull with C-FIND, N-ACTION - Change UPS State, N-GET, N-SET primitives from the UPS Pull Service group, or rather create distinct class for each SOP class (where the combination would be handled in another class or directly in the CLI app. As the primitive C-FIND, N-SET and N-GET are present in multiple UPS service groups, one handler class for all DIMSE primitive makes sense. What do you think?

I think we need to consider whether the UPS SCP is intended to be re-usable, in the sense of a base class that can be derived from, or as a factory from which multiple instances would be created in a single process space.  Or if there are parts of the application as it is that we would want to be able to re-use separately.  I'll start thinking about the use cases that would motivate putting in the additional effort to redesign the application, but it's the use cases that I think should drive the design decisions.

**Comment by dwikler** [comment]
> 
> I think we need to consider whether the UPS SCP is intended to be re-usable, in the sense of a base class that can be derived from, or as a factory from which multiple instances would be created in a single process space. Or if there are parts of the application as it is that we would want to be able to re-use separately. I'll start thinking about the use cases that would motivate putting in the additional effort to redesign the application, but it's the use cases that I think should drive the design decisions.

Right, this is a good point. I think there may be use cases for multiple treatment delivery workflow applications implementing UPS SOP Classes. Concept profiles extending TDW-II with an additional actor, the TSM, is an example where the TSM would implement a local worklist server to manage the steps of the session performed by multiple actors. Another example could be a secondary TMS local to the TDD or specific to the TDD as discussed in the SMRT concept profile. 
Nevertheless, we can leave this enhancement for later and I will start by refactoring upsscp as a CLI app with the only hcange to reuse the CEchoSCP class.

**Comment by dwikler** [comment]
A new pull request (https://github.com/sjswerdloff/tdwii_plus_examples/pull/92) refactors upsscp to use CEchoSCP and refactors config files locations

**Comment by dwikler** [comment]
@sjswerdloff 
All the "SCP" classes and applications have now been refactored. 
- All applications (CLI and UI) rely on the SCP classes.
- All "SCP" classes are now modules only in the tdwii_plus_examples package folder
- All "SCP" CLI apps are now in the cli folder in the package folder.
- All default config files are now in the config file in the package folder

The next step is to refactor remaining modules and apps to have "SCU" classes to support each DIMSE request in CC.2. DIMSE Service Groups
- UPS operation request
  - Create a Unified Procedure Step (N-CREATE)
  - Change UPS State (N-ACTION)
  - Request UPS Cancel (N-ACTION)
  - Subscribe/Unsubscribe to Receive UPS Event Reports (N-ACTION)
  - Search for Unified Procedure Step (C-FIND) 
  - Get Unified Procedure Step Information (N-GET)
  - Set Unified Procedure Step Information (N-SET)
- UPS notification request
  - Report a Change in UPS Status (N-EVENT-REPORT)
 
One way is to have one class per DIMSE Service Group (or per UPS Sop Class) which would however duplicate some code for the N-GET, C-FIND, N-ACTION - Request UPS Cancel which are present in more than one SOP Class
- UPSPushSCU
- UPSPullSCU
- UPSWatchSCU (already exist)
- UPSEventSCU

Another way is to have one class per DIMSE but this would bring DIMSE from different UPS SOP classes together
- UPSNActionSCU
- UPSCFindSCU
- UPSNGetSCU
- UPSNSetSCU
- UPSNEventSender

Or we could have one class per operation/notification
- UPSCreateSCU 
- UPSCancelSCU
- UPSChangeStateSCU
- UPSearchSCU
- UPSGetSCU
- UPSSetSCU
- UPSWatchSCU (already exist)
- UPSNEventSender

Any suggestions?

**Comment by sjswerdloff** [comment]
Python supports mixin classes (multiple inheritance).  I think one class per DIMSE and then various actors would inherit from whichever DIMSE classes they need, and that would be "pythonic".  It also would "read" well.
A particular actor "is a UPSNactionSCU and is a UPSNSetSCU".
Is there something I'm missing that would make that approach less than optimal?

**Comment by dwikler** [comment]
The issue I see is that a subscriber which is not a performer does not need the N-ACTION Type ID 1 which it would get if inheriting from the UPSNactionSCU

**Comment by dwikler** [comment]
## Status of the refactoring 


[Supported_M]: https://img.shields.io/badge/M-32CD32?style=flat-square&logo=check&logoColor=white
[UnSupported_M]: https://img.shields.io/badge/M-8B0000?style=flat-square&logo=check&logoColor=white
[Supported_U]: https://img.shields.io/badge/U-2CD32?style=flat-square&logo=times&logoColor=white
[Unsupported_U]: https://img.shields.io/badge/U-8B0000?style=flat-square&logo=times&logoColor=white
[Unsupported_Ma]: https://img.shields.io/badge/M-FF4500?style=flat-square&logo=times&logoColor=white
[Unsupported_Ua]: https://img.shields.io/badge/U-FF4500?style=flat-square&logo=times&logoColor=white

[Green]: https://img.shields.io/badge/-2CD32?style=flat-square&logo=times&logoColor=white
[Red]: https://img.shields.io/badge/-8B0000?style=flat-square&logo=times&logoColor=white
[Orange]: https://img.shields.io/badge/-FF4500?style=flat-square&logo=times&logoColor=white

| Color Code   | Meaning        
|--------------|------- 
![Green]       | Supported and Refactored
![Red]         | Unsupported
![Orange]      | Implicitly Supported and Not Refactored

Note: _Implicitly Supported_ means that the original implementation does provide the function by including all UPS SOP Classes in association where the refactoring tries to explicitly specified the SOP Class the association.

---

#### UPS Push SOP Class

| DICOM Message Service Element | SCU Usage            | SCP Usage            |
|------------------------------ |--------------------- |--------------------- |
| N-CREATE                      | ![Supported_M]       | ![Supported_M]       |
| N-ACTION - Request UPS Cancel | ![Unsupported_U]     | ![Unsupported_M]       |
| N-GET                         | ![Unsupported_Ua]     | ![Supported_M]       |

#### UPS Pull SOP Class

| DICOM Message Service Element | SCU Usage            | SCP Usage            |
|------------------------------ |--------------------- |--------------------- |
| C-FIND                        | ![Supported_M]       | ![Supported_M]       |
| N-GET                         | ![Unsupported_Ua]     | ![Supported_M]       |
| N-SET                         | ![Supported_M]       | ![Supported_M]       |
| N-ACTION - Change UPS State   | ![Supported_M]       | ![Supported_M]       |

#### UPS Watch SOP Class

| DICOM Message Service Element | SCU Usage            | SCP Usage            |
|------------------------------ |--------------------- |--------------------- |
| N-ACTION - Un/Subscribe       | ![Supported_M]       | ![Supported_M]       |
| N-GET                         | ![Unsupported_Ua]     | ![Supported_M]       |
| C-FIND                        | ![Unsupported_Ua]     | ![Supported_M]       |
| N-ACTION - Request UPS Cancel | ![Unsupported_U]     | ![Unsupported_M]       |

#### UPS Event SOP Class

| DICOM Message Service Element | SCU Usage            | SCP Usage            |
|------------------------------ |--------------------- |--------------------- |
| N-EVENT-REPORT                | ![Supported_M]       | ![Supported_M]       |

#### UPS Query SOP Class

| DICOM Message Service Element | SCU Usage            | SCP Usage            |
|------------------------------ |--------------------- |--------------------- |
| C-FIND                        | ![Unsupported_Ma]     | ![Unsupported_Ma]     |

**Comment by sjswerdloff** [comment]
> Before closing the issue, we could perform the following tasks:
> 
> * Clean up obsolete modules files
> * Refactor UPS Pull N-GET (not used in IHE-RO TDW-II)

I think it's very important to clean up/remove the obsoleted modules/files.  Otherwise people might use them and run in to problems.

I think we should create a separate issue for refactoring the UPS Pull N-GET.
It isn't critical to TDW-II nor preliminary use of events, and not likely to see immediate use, but it would be good in terms of eventually implementing the majority of UPS (DIMSE), and might be valuable in potential newer TDW versions (TDW-III, TDW-Web (for proxy-adapter purposes))

**Comment by sjswerdloff** [comment]
We also need to update the README.md to point to the cli commands.
